compiler/rustc_llvm/build.rs

```diff
@@ -258,6 +258,12 @@ fn main() {
     {
         println!("cargo:rustc-link-lib=z");
     } else if target.contains("netbsd") {
+        // Building for i586 or i686, we need -latomic for 64-bit atomics
```

Comment:
Does that use lock-free atomics? Also is this needed for the rust standard library as well?

Reply:
To be perfectly honest, I have not looked under the hood for that.  The `-latomic` library comes from `gcc`.  I believe this is the only available solution for this platform since LLVM stubbornly insists on relying on 64-bit atomic operations.  i586, which is "Pentium"-class CPUs, do, as far as I'm aware, not have any native-to-the-CPU 64-bit atomic instructions, so I cling on to what appears to be the only reasonable alternative.

As for whether this is also needed for the rust standard library?  I don't *think* so -- I think these were the only i586-specific patches that I could find in our `pkgsrc` package (though I still need to submit pull requests for some of the vendored components, such as the rust `libc`).  The resulting rust compiler has been used to cross-build and build rust for this target for quite a while, and the resulting compiler has successfully been used to build e.g. firefox (natively).

Reply:
If the cpu doesn't support 64bit atomics, `max_atomic_width` should probably be 32bit. We guarantee that atomics are lock-free and declare them as unavailable when they wouldn't be. We don't emulate them using mutexes.

Reply:
That's a fair point as far as rust itself goes, and I see that the target spec says 64, so this sends me back to the test bench to see if the build will succeed with that set to 32 instead.

However, that said, the above modification concerns LLVM's reliance on 64-bit atomics, and that also needs to be dealt with one way or the other.

Reply:
That said, the `i586_pc_nto_qnx700.rs` target spec both says `llvm_target: "i586-pc-unknown".into(),` and sets atomic width to 64 bits.  How it manages to do that I do not know...

Reply:
I changed `max_atomic_width` to 32 and re-tried the build of `rust` from scratch.

The initial attempt fell on its face with

```
     Running `/usr/pkgsrc/wip/rust/work/rustc-1.73.0-src/build/bootstrap/debug/rustc --crate-name crossbeam_utils --edition=2018 /usr/pkgsrc/wip/rust/work/rustc-1.73.0-src/vendor/crossbeam-utils/src/lib.rs --error-format=json --json=diagnostic-rendered-ansi,artifacts,future-incompat --crate-type lib --emit=dep-info,metadata,link -C opt-level=3 -C embed-bitcode=no --cfg 'feature="std"' -Zunstable-options --check-cfg 'values(feature, "default", "loom", "nightly", "std")' --check-cfg 'names()' --check-cfg 'values()' -C metadata=e0b5eeecf1f9537e -C extra-filename=-e0b5eeecf1f9537e --out-dir /usr/pkgsrc/wip/rust/work/rustc-1.73.0-src/build/i586-unknown-netbsd/stage1-rustc/i586-unknown-netbsd/release/deps --target i586-unknown-netbsd -L dependency=/usr/pkgsrc/wip/rust/work/rustc-1.73.0-src/build/i586-unknown-netbsd/stage1-rustc/i586-unknown-netbsd/release/deps -L dependency=/usr/pkgsrc/wip/rust/work/rustc-1.73.0-src/build/i586-unknown-netbsd/stage1-rustc/release/deps --extern cfg_if=/usr/pkgsrc/wip/rust/work/rustc-1.73.0-src/build/i586-unknown-netbsd/stage1-rustc/i586-unknown-netbsd/release/deps/libcfg_if-6e2d2163a7e2c47b.rmeta --cap-lints allow --cfg=windows_raw_dylib -Csymbol-mangling-version=v0 -Zunstable-options '--check-cfg=values(bootstrap)' '--check-cfg=values(parallel_compiler)' '--check-cfg=values(no_btreemap_remove_entry)' '--check-cfg=values(crossbeam_loom)' '--check-cfg=values(span_locations)' '--check-cfg=values(rustix_use_libc)' '--check-cfg=values(emulate_second_only_system)' '--check-cfg=values(windows_raw_dylib)' -Zmacro-backtrace -Clink-args=-Wl,-z,origin -Clink-args=-Wl,-rpath,/usr/pkg/lib -Zunstable-options -Csplit-debuginfo=off -Zunstable-options '-Wrustc::internal' -Cprefer-dynamic -Z binary-dep-depinfo`
error[E0432]: unresolved imports `core::sync::atomic::AtomicI64`, `core::sync::atomic::AtomicU64`
  --> /usr/pkgsrc/wip/rust/work/rustc-1.73.0-src/vendor/crossbeam-utils/src/lib.rs:79:49
   |
79 |             pub(crate) use core::sync::atomic::{AtomicI64, AtomicU64};
   |                                                 ^^^^^^^^^  ^^^^^^^^^ no `AtomicU64` in `sync::atomic`
   |                                                 |
   |                                                 no `AtomicI64` in `sync::atomic`
   |
help: a similar name exists in the module
   |
79 |             pub(crate) use core::sync::atomic::{AtomicI32, AtomicU64};
   |                                                 ~~~~~~~~~
help: a similar name exists in the module
   |
79 |             pub(crate) use core::sync::atomic::{AtomicI64, AtomicU32};
   |                                                            ~~~~~~~~~

error[E0412]: cannot find type `AtomicU64` in module `core::sync::atomic`
    --> /usr/pkgsrc/wip/rust/work/rustc-1.73.0-src/vendor/crossbeam-utils/src/atomic/consume.rs:78:14
     |
78   | impl_atomic!(AtomicU64, u64);
     |              ^^^^^^^^^ help: a struct with a similar name exists: `AtomicU16`
     |
    ::: /usr/pkgsrc/wip/rust/work/rustc-1.73.0-src/library/core/src/sync/atomic.rs:2031:9
     |
2031 |         pub struct $atomic_type {
     |         ----------------------- similarly named struct `AtomicU16` defined here

error[E0412]: cannot find type `AtomicI64` in module `core::sync::atomic`
    --> /usr/pkgsrc/wip/rust/work/rustc-1.73.0-src/vendor/crossbeam-utils/src/atomic/consume.rs:80:14
     |
80   | impl_atomic!(AtomicI64, i64);
     |              ^^^^^^^^^ help: a struct with a similar name exists: `AtomicI16`
     |
    ::: /usr/pkgsrc/wip/rust/work/rustc-1.73.0-src/library/core/src/sync/atomic.rs:2031:9
     |
2031 |         pub struct $atomic_type {
     |         ----------------------- similarly named struct `AtomicI16` defined here

Some errors have detailed explanations: E0412, E0432.
For more information about an error, try `rustc --explain E0412`.

Did not run successfully: exit status: 1
```

However, with this change to the vendored `crossbeam-utils` crate(s), I get further along:

```
--- vendor/crossbeam-utils/no_atomic.rs.orig    2023-10-03 04:12:09.000000000 +0000
+++ vendor/crossbeam-utils/no_atomic.rs
@@ -28,15 +28,16 @@ const NO_ATOMIC_64: &[&str] = &[
     "armv6k-nintendo-3ds",
     "avr-unknown-gnu-atmega328",
     "hexagon-unknown-linux-musl",
+    "i586-unknown-netbsd",
     "m68k-unknown-linux-gnu",
     "mips-unknown-linux-gnu",
     "mips-unknown-linux-musl",
```

(Though, mysteriosly, my first attempt ended up with a SEGV, but re-starting the build is still progressing beyond that point.) ... But ... the build still fails with

```
error[E0412]: cannot find type `AtomicU64` in module `atomic`
    --> /usr/pkgsrc/wip/rust/work/rustc-1.73.0-src/vendor/zip/src/types.rs:282:30
     |
282  | pub struct AtomicU64(atomic::AtomicU64);
     |                              ^^^^^^^^^ help: a struct with a similar name exists: `AtomicU16`
     |
    ::: /usr/pkgsrc/wip/rust/work/rustc-1.73.0-src/library/core/src/sync/atomic.rs:2031:9
     |
2031 |         pub struct $atomic_type {
     |         ----------------------- similarly named struct `AtomicU16` defined here
     |
note: struct `crate::write::zip_writer::AtomicU64` exists but is inaccessible
    --> /usr/pkgsrc/wip/rust/work/rustc-1.73.0-src/vendor/zip/src/types.rs:282:1
     |
282  | pub struct AtomicU64(atomic::AtomicU64);
     | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ not accessible

error[E0433]: failed to resolve: could not find `AtomicU64` in `atomic`
   --> /usr/pkgsrc/wip/rust/work/rustc-1.73.0-src/vendor/zip/src/types.rs:286:22
    |
286 |         Self(atomic::AtomicU64::new(v))
    |                      ^^^^^^^^^
    |                      |
    |                      could not find `AtomicU64` in `atomic`
    |                      help: a struct with a similar name exists: `AtomicU16`
    |
note: struct `crate::write::zip_writer::AtomicU64` exists but is inaccessible
   --> /usr/pkgsrc/wip/rust/work/rustc-1.73.0-src/vendor/zip/src/types.rs:282:1
    |
282 | pub struct AtomicU64(atomic::AtomicU64);
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ not accessible

error[E0433]: failed to resolve: could not find `AtomicU64` in `atomic`
   --> /usr/pkgsrc/wip/rust/work/rustc-1.73.0-src/vendor/zip/src/types.rs:304:22
    |
304 |         Self(atomic::AtomicU64::new(self.load()))
    |                      ^^^^^^^^^
    |                      |
    |                      could not find `AtomicU64` in `atomic`
    |                      help: a struct with a similar name exists: `AtomicU16`
    |
note: struct `crate::write::zip_writer::AtomicU64` exists but is inaccessible
   --> /usr/pkgsrc/wip/rust/work/rustc-1.73.0-src/vendor/zip/src/types.rs:282:1
    |
282 | pub struct AtomicU64(atomic::AtomicU64);
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ not accessible

Some errors have detailed explanations: E0412, E0433.
For more information about an error, try `rustc --explain E0412`.
```

and I don't see any place where that use can be easily avoided.  So ... This goes back to "rust or its included components really, *really* want to have access to 64-bit atomics".

Reply:
> However, that said, the above modification concerns LLVM's reliance on 64-bit atomics, and that also needs to be dealt with one way or the other.

For that `-latomic` is fine.

> and I don't see any place where that use can be easily avoided. So ... This goes back to "rust or its included components really, really want to have access to 64-bit atomics".

Looks like those crates are hard-coding which targets lack 64-bit atomics, which used to be necessary as there was no stable way to query which atomic widths are supported, but rust 1.60.0 stabilized the `target_has_atomic` family of `#[cfg]`'s which allow querying this at compile time.

Also if I try to use `AtomicU64::compare_exchange` on i586-unknown-linux-gnu, I get an `lock cmpxchg8b`, not a call to libatomic which suggests that 64-bit atomics are supported on i586 after all.

Reply:
OK, it looks like you are correct wrt. `cmpxchg8b` -- it's there from Pentium onwards.  So...  I assume it's then OK to let atomic size remain at 64 for this target?  Perhaps the comment related to `-latomic` should be updated to say why it's there -- I *think* it's there because NetBSD "i386" family compilers `gcc` and `g++` are both built for i486 (under the motto "we care about backward compatibility", I'm sure), and LLVM wants/needs 64-bit atomics, ref.:

```
: {1} gcc -v
Using built-in specs.
COLLECT_GCC=gcc
COLLECT_LTO_WRAPPER=/usr/libexec/lto-wrapper
Target: i486--netbsdelf
Configured with: /usr/src/tools/gcc/../../external/gpl3/gcc/dist/configure --target=i486--netbsdelf --enable-long-long --enable-threads --with-bugurl=http://www.NetBSD.org/Misc/send-pr.html --with-pkgversion='NetBSD nb4 20200810' --with-system-zlib --without-isl --enable-__cxa_atexit --enable-libstdcxx-time=rt --enable-libstdcxx-threads --with-diagnostics-color=auto-if-env --with-arch=i486 --with-tune=nocona --with-default-libstdcxx-abi=new --with-mpc-lib=/var/obj/mknative/i386/usr/src/external/lgpl3/mpc/lib/libmpc --with-mpfr-lib=/var/obj/mknative/i386/usr/src/external/lgpl3/mpfr/lib/libmpfr --with-gmp-lib=/var/obj/mknative/i386/usr/src/external/lgpl3/gmp/lib/libgmp --with-mpc-include=/usr/src/external/lgpl3/mpc/dist/src --with-mpfr-include=/usr/src/external/lgpl3/mpfr/dist/src --with-gmp-include=/usr/src/external/lgpl3/gmp/lib/libgmp/arch/i386 --enable-tls --disable-multilib --disable-libstdcxx-pch --build=i486--netbsdelf --host=i486--netbsdelf --with-sysroot=/var/obj/mknative/i386/usr/src/destdir.i386
Thread model: posix
gcc version 7.5.0 (nb4 20200810) 
: {2} 
: {2} uname -a
NetBSD xxxx.uninett.no 9.2 NetBSD 9.2 (GENERIC_PAE) #0: Sat Apr 16 13:06:03 CEST 2022  he@xxxx.uninett.no:/usr/obj/sys/arch/i386/compile/GENERIC_PAE i386
: {3} 
```

Reply:
> I assume it's then OK to let atomic size remain at 64 for this target?

Looks like it. I opened an issue for crossbeam and zip to improve compatibility with other targets without 64-bit atomics, but that isn't necessary for i586-unknown-netbsd it seems.

> I think it's there because NetBSD "i386" family compilers gcc and g++ are both built for i486

That would explain it.

Reply:
> > I think it's there because NetBSD "i386" family compilers gcc and g++ are both built for i486
> 
> That would explain it.

Could you add a comment to that effect?

Reply:
> > > I think it's there because NetBSD "i386" family compilers gcc and g++ are both built for i486
> >
> > That would explain it.
>
> Could you add a comment to that effect?

Done.